services/venice-controller/src/main/java/com/linkedin/venice/controller/UnusedValueSchemaCleanupService.java

```diff
@@ -0,0 +1,127 @@
+package com.linkedin.venice.controller;
+
+import com.linkedin.venice.common.VeniceSystemStoreUtils;
+import com.linkedin.venice.meta.Store;
+import com.linkedin.venice.meta.Version;
+import com.linkedin.venice.schema.SchemaEntry;
+import com.linkedin.venice.service.AbstractVeniceService;
+import com.linkedin.venice.utils.DaemonThreadFactory;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+
+public class UnusedValueSchemaCleanupService extends AbstractVeniceService {
+  private final ScheduledExecutorService executor =
+      Executors.newSingleThreadScheduledExecutor(new DaemonThreadFactory("UnusedValueSchemaCleanupService"));
+  private final VeniceControllerMultiClusterConfig multiClusterConfig;
+  private final VeniceParentHelixAdmin veniceParentHelixAdmin;
+  private final VeniceHelixAdmin admin;
+  private final int scheduleIntervalMinutes;
+  private boolean stop = false;
+  public final int minSchemaCountToKeep;
+
+  UnusedValueSchemaCleanupService(
+      VeniceControllerMultiClusterConfig multiClusterConfig,
+      VeniceHelixAdmin admin,
+      VeniceParentHelixAdmin parentHelixAdmin) {
+    this.multiClusterConfig = multiClusterConfig;
+    this.admin = admin;
+    this.scheduleIntervalMinutes = multiClusterConfig.getUnusedSchemaCleanupIntervalMinutes();
+    this.minSchemaCountToKeep = multiClusterConfig.getMinSchemaCountToKeep();
+    this.veniceParentHelixAdmin = parentHelixAdmin;
+  }
+
+  private Runnable getRunnableForSchemaCleanup() {
+    return () -> {
+      if (stop) {
+        return;
+      }
+      for (String clusterName: multiClusterConfig.getClusters()) {
+        boolean cleanupEnabled =
+            multiClusterConfig.getControllerConfig(clusterName).isUnusedValueSchemaCleanupServiceEnabled();
+        if (!cleanupEnabled) {
+          continue;
+        }
+        // Get all stores for current cluster
+        List<Store> stores = admin.getAllStores(clusterName);
+        for (Store store: stores) {
+          String storeName = store.getName();
+          // Remove schema only for batch stores
+          if (isHybridStore(store) || VeniceSystemStoreUtils.isSystemStore(storeName)) {
+            continue;
+          }
+          Collection<SchemaEntry> allSchemas = veniceParentHelixAdmin.getValueSchemas(clusterName, storeName);
+
+          if (allSchemas.size() < minSchemaCountToKeep) {
+            continue;
+          }
+
+          Set<Integer> usedSchemaSet = veniceParentHelixAdmin.getInUseValueSchemaIds(clusterName, storeName);
+
+          if (usedSchemaSet.isEmpty()) {
```

Comment:
Can we also have an unit test to make sure in the case where we are unable to `getInUseValueSchemaIds` due to errors or unreachable child colo, the schema deletion is skipped just like what you implemented here? (This is to make sure this behavior stays when some of these APIs get modified in the future to be used for other things as well)

Reply:
Same here, a simple unit test will do.